File: src/lib/contracts.ts
export type LearnerLevel = "beginner" | "intermediate" | "advanced";
export type SubtitleDisplayMode = "source" | "translation" | "both";
export type CaptionProcessingMode = "translated" | "original_only";
export type SegmentStatus = "pending" | "complete" | "failed" | "skipped";
export type SessionMode = "file" | "live";
export type LiveSyncMode = "coordinated" | "fast_source";
export type LiveTranslationEngine = "realtime" | "transcript_locked";
export type LiveSyncStatus = "steady" | "catching_up" | "degraded";
export type LiveCaptureLifecycle = "inactive" | "starting" | "active" | "reconnecting" | "stopping" | "failed";
export type EndSessionReason = "user_stop" | "replacement" | "quit" | "fatal_error";
export type SessionPhase =
  | "idle"
  | "preparing"
  | "transcribing"
  | "buffering"
  | "ready"
  | "playing"
  | "paused"
  | "reconnecting"
  | "complete";

export type SubtitlePreset =
  | "clean"
  | "classic-outline"
  | "yellow-drop"
  | "arcade"
  | "momento"
  | "wired";
export type AppSurface = "workbench" | "viewer" | "overlay" | "lesson" | "launcher";
export type LauncherMode = "file" | "live";
export type LauncherState = "idle" | "hovering" | "preparing" | "starting" | "error";
export type LiveCaptureSourceKind = "application" | "window" | "display";

export interface LiveCaptureSourceSelection {
  kind: LiveCaptureSourceKind;
  processId?: number;
  windowId?: number;
  displayId?: number;
}

export interface LiveCaptureSource extends LiveCaptureSourceSelection {
  id: string;
  title: string;
  detail: string;
  applicationName?: string;
  bundleIdentifier?: string;
}

export interface LiveCaptureSources {
  applications: LiveCaptureSource[];
  windows: LiveCaptureSource[];
  displays: LiveCaptureSource[];
}

export interface LessonPlacement {
  monitorKey: string;
  x: number;
  y: number;
}

export type LessonSurfaceMode = "compose" | "thinking" | "lesson" | "error";

export interface LessonOpenContext {
  selectionId: number;
  sessionId: string;
  sourceSurface: "viewer" | "overlay" | "workbench";
  segmentId: string;
  selectedSegment: SubtitleSegment;
  cursorX: number;
  cursorY: number;
  externalMediaControl: ExternalMediaControlResult;
}

export interface LessonClosedContext {
  selectionId: number;
  sessionId: string;
  sourceSurface: "viewer" | "overlay" | "workbench";
  segmentId: string;
  reason: "closed" | "invalidated";
}

export type ExternalMediaControlResult = "not_requested" | "paused" | "permission_required" | "failed" | "unsupported";

export interface LanguageSettings {
  source: "auto" | string;
  target: string;
  explanation: string;
}

export interface SyncSettings {
  liveMode: LiveSyncMode;
  translationEngine: LiveTranslationEngine;
}

export interface LiveSyncState {
  targetDelayMs: number;
  observedLagMs: number;
  status: LiveSyncStatus;
  visibleSegmentId?: string;
}

export interface LiveCaptureStatus {
  sessionId: string;
  lifecycle: LiveCaptureLifecycle;
  startedAtMs?: number;
  sourceLabel?: string;
}

export interface SessionEnding {
  sessionId: string;
  reason: EndSessionReason;
}

export interface SubtitleSegment {
  id: string;
  origin: SessionMode;
  startMs: number;
  endMs: number;
  sourceText: string;
  translationText?: string;
  ambiguityNote?: string;
  speakerId?: string;
  isProvisional: boolean;
  transcriptionStatus: SegmentStatus;
  translationStatus: SegmentStatus;
}

export interface SpeakerProfile {
  id: string;
  displayName: string;
  color: string;
  reference?: { startMs: number; endMs: number };
}

export interface WiredColors {
  panel: string;
  wash: string;
  sourceText: string;
  translationText: string;
  metadata: string;
  fallbackAccent: string;
}

export interface ArcadeColors {
  text: string;
  panel: string;
}

export interface StyleSettings {
  preset: SubtitlePreset;
  position: { x: number; y: number };
  overlayPosition: { x: number; y: number };
  overlayWidth: number;
  fontFamily: string;
  fontSize: number;
  backgroundOpacity: number;
  effect: "none" | "outline" | "shadow";
  displayMode: SubtitleDisplayMode;
  showSpeakerNames: boolean;
  wiredColors: WiredColors;
  arcadeColors: ArcadeColors;
}

export interface BoardSection {
  heading: string;
  lines: ChalkPhrase[];
}

export type BoardDemoKind = "none" | "sentence_breakdown" | "omitted_meaning" | "literal_to_natural" | "tone_scale" | "mini_dialogue";
export type ChalkColor = "white" | "baby_blue" | "yellow" | "pink";
export type ChalkMark = "none" | "box" | "bracket" | "strike";
export type TailCue = "none" | "point" | "underline";

export interface ChalkPhrase {
  text: string;
  color: ChalkColor;
  mark: ChalkMark;
  tailCue: TailCue;
}

export interface SourceFocus {
  color: ChalkColor;
  tailCue: TailCue;
}

export interface BoardDemoItem {
  label: string;
  detail: string;
  color: ChalkColor;
  mark: ChalkMark;
  tailCue: TailCue;
}

export interface BoardDemo {
  kind: BoardDemoKind;
  caption?: string;
  items: BoardDemoItem[];
  result?: ChalkPhrase;
}

export interface TeachingMoment {
  title: string;
  speechBubble: string;
  sourceFocus: SourceFocus;
  boardSections: BoardSection[];
  demonstration: BoardDemo;
  ambiguityNote?: ChalkPhrase;
}

export interface LessonCard {
  schemaVersion: 2;
  selectedSegmentId: string;
  moments: TeachingMoment[];
  suggestedFollowUps: string[];
}

export interface LessonMessage {
  id: string;
  role: "user" | "assistant";
  text: string;
  card?: LessonCard;
}

export interface RecoverableError {
  code: string;
  message: string;
  segmentId?: string;
}

export interface RetranslatedSegment {
  segmentId: string;
  translationText: string;
  ambiguityNote?: string;
}

export type SessionEvent =
  | { type: "session_reset"; mode: SessionMode; languages: LanguageSettings; processingMode: CaptionProcessingMode; liveTranslationEngine?: LiveTranslationEngine }
  | { type: "phase_changed"; phase: SessionPhase }
  | { type: "caption_upserted"; segment: SubtitleSegment }
  | { type: "transcript_finalized"; segment: SubtitleSegment }
  | { type: "translation_finalized"; segmentId: string; translationText: string; ambiguityNote?: string }
  | { type: "file_retranslation_applied"; languages: LanguageSettings; translations: RetranslatedSegment[] }
  | { type: "speaker_discovered"; speaker: SpeakerProfile }
  | { type: "coverage_changed"; readyThroughMs: number }
  | { type: "live_sync_changed"; sync: LiveSyncState }
  | { type: "live_audio_gap"; startMs: number; endMs: number }
  | { type: "lesson_selected"; segmentId?: string }
  | { type: "recoverable_error"; error: RecoverableError }
  | { type: "fatal_error"; message: string }
  | { type: "complete" };

export interface SequencedSessionEvent {
  sessionId: string;
  sequence: number;
  event: SessionEvent;
}

export interface PreparedMediaInfo {
  path: string;
  fileName: string;
}

export interface SessionState {
  sessionId: string;
  sequence: number;
  mode?: SessionMode;
  processingMode: CaptionProcessingMode;
  liveTranslationEngine?: LiveTranslationEngine;
  languages: LanguageSettings;
  phase: SessionPhase;
  segments: SubtitleSegment[];
  speakers: Record<string, SpeakerProfile>;
  readyThroughMs: number;
  liveSync?: LiveSyncState;
  errors: RecoverableError[];
  fatalError?: string;
  selectedSegmentId?: string;
  media?: PreparedMediaInfo;
}

export type CapabilityAvailability = "available" | "unavailable" | "unknown";

export interface ApiConfigurationStatus {
  configured: boolean;
  validatedAt?: number;
  validationSchema: number;
  languageModel: CapabilityAvailability;
  fileTranscription: CapabilityAvailability;
  realtimeTranslation: CapabilityAvailability;
  realtimeOriginalOnly: CapabilityAvailability;
  liveTextTranslation: CapabilityAvailability;
}

/** @deprecated Use ApiConfigurationStatus. */
export interface ModelReadiness {
  file: boolean;
  live: boolean;
}

export const DEFAULT_LANGUAGES: LanguageSettings = {
  source: "auto",
  target: "en",
  explanation: "en",
};

export const DEFAULT_SYNC: SyncSettings = {
  liveMode: "coordinated",
  translationEngine: "realtime",
};

export const DEFAULT_LIVE_SYNC: LiveSyncState = {
  targetDelayMs: 2_500,
  observedLagMs: 0,
  status: "steady",
};

export const DEFAULT_STYLE: StyleSettings = {
  preset: "momento",
  position: { x: 0.5, y: 0.82 },
  overlayPosition: { x: 0.5, y: 0.78 },
  overlayWidth: 900,
  fontFamily: "Inter",
  fontSize: 28,
  backgroundOpacity: 0.58,
  effect: "outline",
  displayMode: "both",
  showSpeakerNames: true,
  wiredColors: {
    panel: "#05081c",
    wash: "#0b2944",
    sourceText: "#c9e6fa",
    translationText: "#ffffff",
    metadata: "#5fa8dc",
    fallbackAccent: "#4ac8ff",
  },
  arcadeColors: {
    text: "#f0a14a",
    panel: "#0b0d08",
  },
};

export const EMPTY_SESSION: SessionState = {
  sessionId: "fixture",
  sequence: 0,
  processingMode: "translated",
  languages: { ...DEFAULT_LANGUAGES },
  phase: "idle",
  segments: [],
  speakers: {},
  readyThroughMs: 0,
  errors: [],
};

File: src-tauri/src/contracts.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum LearnerLevel {
    Beginner,
    Intermediate,
    Advanced,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SegmentStatus {
    Pending,
    Complete,
    Failed,
    Skipped,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CaptionProcessingMode {
    #[default]
    Translated,
    OriginalOnly,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SessionMode {
    File,
    Live,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum LiveSyncMode {
    #[default]
    Coordinated,
    FastSource,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum LiveTranslationEngine {
    #[default]
    Realtime,
    TranscriptLocked,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum LiveSyncStatus {
    Steady,
    CatchingUp,
    Degraded,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LiveSyncState {
    pub target_delay_ms: u64,
    pub observed_lag_ms: u64,
    pub status: LiveSyncStatus,
    pub visible_segment_id: Option<String>,
}

impl Default for LiveSyncState {
    fn default() -> Self {
        Self {
            target_delay_ms: 2_500,
            observed_lag_ms: 0,
            status: LiveSyncStatus::Steady,
            visible_segment_id: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LanguageSettings {
    pub source: String,
    pub target: String,
    pub explanation: String,
}

impl Default for LanguageSettings {
    fn default() -> Self {
        Self {
            source: "auto".into(),
            target: "en".into(),
            explanation: "en".into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleSegment {
    pub id: String,
    pub origin: SessionMode,
    pub start_ms: u64,
    pub end_ms: u64,
    pub source_text: String,
    pub translation_text: Option<String>,
    pub ambiguity_note: Option<String>,
    pub speaker_id: Option<String>,
    pub is_provisional: bool,
    pub transcription_status: SegmentStatus,
    pub translation_status: SegmentStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SpeakerProfile {
    pub id: String,
    pub display_name: String,
    pub color: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RecoverableError {
    pub code: String,
    pub message: String,
    pub segment_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RetranslatedSegment {
    pub segment_id: String,
    pub translation_text: String,
    pub ambiguity_note: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum SessionEvent {
    SessionReset {
        mode: SessionMode,
        languages: LanguageSettings,
        processing_mode: CaptionProcessingMode,
        live_translation_engine: Option<LiveTranslationEngine>,
    },
    PhaseChanged {
        phase: String,
    },
    CaptionUpserted {
        segment: SubtitleSegment,
    },
    TranscriptFinalized {
        segment: SubtitleSegment,
    },
    TranslationFinalized {
        segment_id: String,
        translation_text: String,
        ambiguity_note: Option<String>,
    },
    FileRetranslationApplied {
        languages: LanguageSettings,
        translations: Vec<RetranslatedSegment>,
    },
    SpeakerDiscovered {
        speaker: SpeakerProfile,
    },
    CoverageChanged {
        ready_through_ms: u64,
    },
    LiveSyncChanged {
        sync: LiveSyncState,
    },
    LiveAudioGap {
        start_ms: u64,
        end_ms: u64,
    },
    LessonSelected {
        segment_id: Option<String>,
    },
    RecoverableError {
        error: RecoverableError,
    },
    FatalError {
        message: String,
    },
    Complete,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SequencedSessionEvent {
    pub session_id: String,
    pub sequence: u64,
    pub event: SessionEvent,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PreparedMediaInfo {
    pub path: String,
    pub file_name: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSnapshot {
    pub session_id: String,
    pub sequence: u64,
    pub mode: Option<SessionMode>,
    pub processing_mode: CaptionProcessingMode,
    pub live_translation_engine: Option<LiveTranslationEngine>,
    pub languages: LanguageSettings,
    pub phase: String,
    pub segments: Vec<SubtitleSegment>,
    pub speakers: HashMap<String, SpeakerProfile>,
    pub ready_through_ms: u64,
    pub live_sync: Option<LiveSyncState>,
    pub errors: Vec<RecoverableError>,
    pub fatal_error: Option<String>,
    pub selected_segment_id: Option<String>,
    pub media: Option<PreparedMediaInfo>,
}

impl Default for SessionSnapshot {
    fn default() -> Self {
        Self {
            session_id: "idle".into(),
            sequence: 0,
            mode: None,
            processing_mode: CaptionProcessingMode::Translated,
            live_translation_engine: None,
            languages: LanguageSettings::default(),
            phase: "idle".into(),
            segments: Vec::new(),
            speakers: HashMap::new(),
            ready_through_ms: 0,
            live_sync: None,
            errors: Vec::new(),
            fatal_error: None,
            selected_segment_id: None,
            media: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BoardSection {
    pub heading: String,
    pub lines: Vec<ChalkPhrase>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ChalkColor {
    White,
    BabyBlue,
    Yellow,
    Pink,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ChalkMark {
    None,
    Box,
    Bracket,
    Strike,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TailCue {
    None,
    Point,
    Underline,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ChalkPhrase {
    pub text: String,
    pub color: ChalkColor,
    pub mark: ChalkMark,
    pub tail_cue: TailCue,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceFocus {
    pub color: ChalkColor,
    pub tail_cue: TailCue,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum BoardDemoKind {
    None,
    SentenceBreakdown,
    OmittedMeaning,
    LiteralToNatural,
    ToneScale,
    MiniDialogue,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum NonoGesture {
    #[default]
    Neutral,
    ThumbsUp,
    PointUser,
    PointSelf,
    Cheer,
    HeartTouch,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BoardDemoItem {
    pub label: String,
    pub detail: String,
    pub color: ChalkColor,
    pub mark: ChalkMark,
    pub tail_cue: TailCue,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BoardDemo {
    pub kind: BoardDemoKind,
    pub caption: Option<String>,
    pub items: Vec<BoardDemoItem>,
    pub result: Option<ChalkPhrase>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TeachingMoment {
    pub title: String,
    pub speech_bubble: String,
    #[serde(default)]
    pub gesture: NonoGesture,
    pub source_focus: SourceFocus,
    pub board_sections: Vec<BoardSection>,
    pub demonstration: BoardDemo,
    pub ambiguity_note: Option<ChalkPhrase>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LessonCard {
    pub schema_version: u8,
    pub selected_segment_id: String,
    pub moments: Vec<TeachingMoment>,
    pub suggested_follow_ups: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TutorMessage {
    pub role: String,
    pub text: String,
}
